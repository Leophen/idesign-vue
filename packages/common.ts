import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
