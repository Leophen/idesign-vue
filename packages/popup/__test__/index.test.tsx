import { vi } from 'vitest';
import { Popup } from '../index';
import { render } from '../../common'

describe('Popup 组件测试', () => {
  it('create', () => {
    const wrapper = render(
      <Popup content="提示内容">
        <button>悬浮提示</button>
      </Popup>
    );
    expect(wrapper.element).toMatchSnapshot();
  });

  it('trigger', () => {
    const clickFn = vi.fn();
    const wrapper = render(
      <Popup content="提示内容" trigger="click" onTrigger={clickFn}>
        <button data-testid='trigger'>点击提示</button>
      </Popup>
    );

    wrapper.find('[data-testid="trigger"]').trigger('click');
    expect(clickFn).toBeCalledTimes(1);
  });

  it('disabled', () => {
    const clickFn = vi.fn();
    const wrapper = render(
      <Popup content="提示内容" trigger="click" onTrigger={clickFn} disabled>
        <button data-testid='trigger'>点击提示</button>
      </Popup>
    );

    wrapper.find('[data-testid="trigger"]').trigger('click');
    expect(clickFn).toBeCalledTimes(0);
  });
});

