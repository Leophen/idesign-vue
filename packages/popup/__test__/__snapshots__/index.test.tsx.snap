// Vitest Snapshot v1

exports[`Popup ç»„ä»¶æµ‹è¯• > create 1`] = ``;
